Text index: introduce cache for dictionary blocks

ahmadov opened the pull request:
<!---AI changelog entry and formatting assistance: false-->
### Changelog category (leave one):

- Performance Improvement

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):

Improved text index performance by caching dictionary blocks and using hashtables for token lookups instead of binary search.

### Details

Since dictionary blocks can be loaded multiple times and they can be large in some cases (either dictionary block size is large or strings are longer), caching the deserialized dictionary blocks improves performance.

Additionally, initializes a hashtable to map a token to TokenPostingsInfo which is faster than the binary search over sorted strings in consecutive lookups.

Follow up is adding similar caches for the posting lists and sparse (+bloom filter) indexes.

ahmadov committed:
Text index: introduce cache for dictionary blocks in granule

Additionally, initialize a hashtable to map a token to
`TokenPostingsInfo` which should be faster than the binary search over
sorted strings.

ahmadov committed:
use `CacheBase` instead of `LRUResourceCache`

Since an instance of `MergeTreeIndexGranuleText` is initialized in every
execution, the cache cannot be part of it. Instead, export to its own
class and provide a singleton API.
However, this makes the dictionary block cache a global instance, so it
needs a unique way to identify block in a granule. To do that, we pass
the part full path, index name and index mark to the granule.

ahmadov committed:
make text index dictionary block cache configurable

Also, remove the singleton approach and move text index dictionary block
cache into `Context`.

ahmadov committed:
Fix block key hash value

Use part full path instead of table id to uniquely identify different
parts.

ahmadov committed:
add unit test

rschu1ze commented on tests/queries/0_stateless/02346_text_index_cache.sql:
02346_text_index_dictionary_cache.sql

rschu1ze commented on src/Storages/MergeTree/TextIndexCache.h:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    template <typename LoadFunc>
+    TextIndexDictionaryBlockCacheEntryPtr getOrSet(UInt128 key, LoadFunc && load_func)
```
What is the cache key? A hash of the dictionary block number (0, 1, 2, ...)? In this case, note that there can be multiple text indexes (having the same dict block numbers), we'll need to include the index name, the index expression or the part name to make this globally unique.

ahmadov replied:
yes, I'm using part path, index name, index mark and block id to generate a unique hash key here: https://github.com/ClickHouse/ClickHouse/blob/300c7d8b0140ebb9d97b67be2910f146bef604d1/src/Storages/MergeTree/MergeTreeIndexText.cpp#L422

ahmadov committed:
address pr comments

ahmadov committed:
fix test

rschu1ze committed:
Merge remote-tracking branch 'ClickHouse/master' into ahmadov/dictionary-block-cache

rschu1ze committed:
Minor adjustments

CurtizJ commented on src/Storages/MergeTree/MergeTreeIndexText.cpp:
```diff
@@ -438,20 +419,26 @@ void MergeTreeIndexGranuleText::analyzeDictionary(MergeTreeIndexReaderStream & s
-        compressed_buffer->seek(offset_in_file, 0);
-        auto dictionary_block = deserializeDictionaryBlock(*data_buffer);
+        auto block_key = TextIndexDictionaryBlockCache::hash(state.path_to_data_part, state.index_name, state.index_mark, block_idx);
+        auto dictionary_block = dictionary_block_cache->getOrSet(
```
I think, there should be a way to bypass the cache by query-level setting.

ahmadov committed:
make cache configurable

test with cleared cache

rschu1ze commented:
@ahmadov We will need to be judicious about the memory consumption of the many caches in ClickHouse. We actually don't know at the moment if dictionary deserialization is really a performance bottleneck, and cache entries stick around (there is no mechanism that clears them up if unused). With the current defaults, the cache can grow up to 1 GB - which is significant on the smallest ClickHouse Cloud instance size (8 GB).

Therefore, I propose to disable the cache by default for now. The best option is to set the respective server settings (max byte size, max entries) to 0. Unfortunately, it is complicated to change server settings from a user perspective. So as an alternative, can we default-disable the new setting `use_dict_cache` which you added yesterday?

ahmadov committed:
disable cache by default

- rename `DROP TEXT INDEX CACHE` -> `DROP TEXT INDEX DICTIONARY CACHE`
- add docs

ahmadov committed:
Merge remote-tracking branch 'origin/master' into ahmadov/dictionary-block-cache

ahmadov committed:
update setting def

ahmadov committed:
fix test

ahmadov committed:
try to fix flaky tests

ahmadov committed:
run text index dictionary cache tests sequentially

ahmadov committed:
fix links in docs
